drivers/i2c/i2c_ll_stm32.c

```diff
@@ -0,0 +1,241 @@
+};
+
+static int i2c_stm32_init(struct device *dev)
+{
+	struct device *clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	const struct i2c_stm32_config *cfg = DEV_CFG(dev);
+	struct i2c_stm32_data *data = DEV_DATA(dev);
+	int ret;
+
+	__ASSERT_NO_MSG(clock);
+	clock_control_on(clock, (clock_control_subsys_t *) &cfg->pclken);
+
+	ret = i2c_stm32_runtime_configure(dev, data->dev_config.raw);
+	if (ret < 0) {
+		SYS_LOG_ERR("i2c: failure initializing");
+		return ret;
+	}
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+	k_sem_init(&data->device_sync_sem, 0, UINT_MAX);
+	cfg->irq_config_func(dev);
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_I2C_1
+
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+static void i2c_stm32_irq_config_func_1(struct device *port);
+#endif
+
+static const struct i2c_stm32_config i2c_stm32_cfg_1 = {
+	.i2c = (I2C_TypeDef *) I2C1_BASE,
+	.pclken = {
+		.enr = LL_APB1_GRP1_PERIPH_I2C1,
+		.bus = STM32_CLOCK_BUS_APB1,
+	},
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+	.irq_config_func = i2c_stm32_irq_config_func_1,
+#endif
+};
+
+static struct i2c_stm32_data i2c_stm32_dev_data_1 = {
+	.dev_config.raw = CONFIG_I2C_1_DEFAULT_CFG,
+};
+
+DEVICE_AND_API_INIT(i2c_stm32_1, CONFIG_I2C_1_NAME, &i2c_stm32_init,
+		    &i2c_stm32_dev_data_1, &i2c_stm32_cfg_1,
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &api_funcs);
+
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+static void i2c_stm32_irq_config_func_1(struct device *dev)
+{
+	IRQ_CONNECT(I2C1_EV_IRQn, CONFIG_I2C_1_IRQ_PRI,
+		   stm32_i2c_event_isr, DEVICE_GET(i2c_stm32_1), 0);
+	irq_enable(I2C1_EV_IRQn);
+
+	IRQ_CONNECT(I2C1_ER_IRQn, CONFIG_I2C_1_IRQ_PRI,
+		   stm32_i2c_error_isr, DEVICE_GET(i2c_stm32_1), 0);
+	irq_enable(I2C1_ER_IRQn);
+}
+#endif
+
+#endif /* CONFIG_I2C_1 */
+
+#ifdef CONFIG_I2C_2
+
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+static void i2c_stm32_irq_config_func_2(struct device *port);
+#endif
+
+static const struct i2c_stm32_config i2c_stm32_cfg_2 = {
+	.i2c = (I2C_TypeDef *) I2C2_BASE,
+	.pclken = {
+		.enr = LL_APB1_GRP1_PERIPH_I2C2,
+		.bus = STM32_CLOCK_BUS_APB1,
+	},
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+	.irq_config_func = i2c_stm32_irq_config_func_2,
+#endif
+};
+
+static struct i2c_stm32_data i2c_stm32_dev_data_2 = {
+	.dev_config.raw = CONFIG_I2C_2_DEFAULT_CFG,
+};
+
+DEVICE_AND_API_INIT(i2c_stm32_2, CONFIG_I2C_2_NAME, &i2c_stm32_init,
+		    &i2c_stm32_dev_data_2, &i2c_stm32_cfg_2,
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &api_funcs);
+
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+static void i2c_stm32_irq_config_func_2(struct device *dev)
+{
+	IRQ_CONNECT(I2C2_EV_IRQn, CONFIG_I2C_2_IRQ_PRI,
+		   stm32_i2c_event_isr, DEVICE_GET(i2c_stm32_2), 0);
+	irq_enable(I2C2_EV_IRQn);
+
+	IRQ_CONNECT(I2C2_ER_IRQn, CONFIG_I2C_2_IRQ_PRI,
+		   stm32_i2c_error_isr, DEVICE_GET(i2c_stm32_2), 0);
+	irq_enable(I2C2_ER_IRQn);
+}
+#endif
+
+#endif /* CONFIG_I2C_2 */
+
+#ifdef CONFIG_I2C_3
+
+#ifndef I2C3_BASE
+#error "I2C_3 is not available on the platform that you selected"
```

Comment:
um it clearly flags the reason for the error without having to dig for it; might seem silly but it will probably prevent some email in the mailing list asking about this build error when it happens...